storage/innobase/dict/dict0stats.cc

```diff
@@ -1763,32 +1846,36 @@ dict_stats_scan_page(
 		? page_rec_get_next_non_del_marked<true>
 		: page_rec_get_next_non_del_marked<false>;
 
-	const bool	should_count_external_pages = n_external_pages != NULL;
-
-	if (should_count_external_pages) {
-		*n_external_pages = 0;
+	if (leaf) {
+		leaf_stats->n_external_pages = 0;
 	}
 
 	rec = get_next(page, page_get_infimum_rec(page));
 
 	if (!rec || rec == page_get_supremum_rec(page)) {
-		/* the page is empty or contains only delete-marked records */
-		*n_diff = 0;
+		/* the page is empty or contains only delete-marked
+		records */
+		leaf_stats->n_diff_on_leaf_page = 0;
 		*out_rec = NULL;
 		return(NULL);
 	}
 
 	offsets_rec = rec_get_offsets(rec, index, offsets_rec, n_core,
 				      ULINT_UNDEFINED, &heap);
 
-	if (should_count_external_pages) {
-		*n_external_pages += btr_rec_get_externally_stored_len(
-			rec, offsets_rec);
+	if (leaf) {
+		leaf_stats->n_external_pages +=
+			btr_rec_get_externally_stored_len(
+				rec, offsets_rec);
 	}
 
 	next_rec = get_next(page, rec);
 
-	*n_diff = 1;
+	leaf_stats->n_diff_on_leaf_page = 1;
+
+	if (stats->n_non_null_on_level != nullptr) {
+		leaf_stats->n_non_null_on_leaf_page = 1;
+	}
```

Comment:
Would it do some harm to always initialize or update `leaf_stats->n_non_null_on_leaf_page`? If not, I think that we should do that, to reduce the number of conditional branches that need to be evaluated at runtime.